### PR TITLE
Improve file handling

### PR DESCRIPTION
**Pull request by cthoyt**
The lesson learned for this PR is that you should split the functionality of opening files from the functionality that reads/writes them. Now, the CLI takes care of opening files for read and write, and the functions in the package are only responsible for interacting with files (and not paths).

This is better because click already gives you tools for making nice unix-like CLIs (e.g., give dash and it sends to standard out). It's also better because it makes the code generic to people who might have file-like objects or other sneaky files that are gzipped.

I'm not sure if RDFlib is supporting this out of the box, so I hope there are tests for the code that I touched.

### CODE REVIEW

File: sssom/cli.py [hunk 1]
```diff
@@ -41,6 +42,13 @@
 output_option = click.option(
     "-o", "--output", help="Output file, e.g. a SSSOM tsv file."
 )
+improved_output_option = click.option(
+    "-o",
+    "--output",
+    help="Output file, e.g. a SSSOM tsv file.",
+    type=click.File(mode="w"),
```
Comment (cthoyt) [inline sssom/cli.py:49]: note that you can have click do all of the hard work for opening files for you. below, you can even set stdout as the default. Then, later, you never have to write code that worries about file paths, handling, parsing, etc.

File: sssom/cli.py [hunk 2]
```diff
@@ -315,15 +312,15 @@ def dosql(query: str, inputs: List[str], output: str):
...
 @click.option("-l", "--limit", type=int)
 @click.option("-P", "--prefix", type=click.Tuple([str, str]), multiple=True)
-@output_option
+@improved_output_option
 def sparql(
-    url: str = None,
-    config=None,
-    graph: str = None,
-    limit: int = None,
-    object_labels: bool = None,
```
Comment (cthoyt) [inline sssom/cli.py:324]: Emphatically do not add default values in `click` command line functions - click will always provide all of them

Comment (cthoyt) [reply sssom/cli.py:324]: click options get none by default anyway via the framework

File: sssom/cli.py [hunk 3]
```diff
@@ -425,7 +422,8 @@ def partition(inputs: List[str], output_directory: str):
         # logging.info(f'Example: {cdoc.mapping_set.mappings[0].subject_id}')
         # logging.info(f'Writing to {ofn}. Size={len(cdoc)}')
         msdf = to_mapping_set_dataframe(cdoc)
-        write_table(msdf, ofn)
+        with open(ofn, "w") as file:
+            write_table(msdf, file)
```
Comment (cthoyt) [inline sssom/cli.py:426]: as good 'ol mando says, this is the way.

Better for the function not to have to think about what kind of input this is - if you _must_, you can always write a wrapper around a function that interacts with files to make it interpret a path string in a certain way

File: sssom/cli.py [hunk 4]
```diff
@@ -494,18 +492,15 @@ def crosstab(input: str, output: str, transpose: bool, fields: Tuple):
     ct = pd.crosstab(df[f1], df[f2])
     if transpose:
         ct = ct.transpose()
-    if output is not None:
-        ct.to_csv(output, sep="\t")
-    else:
-        logging.info(ct)
+    ct.to_csv(output, sep="\t")
```
Comment (cthoyt) [inline sssom/cli.py:495]: to_csv accepts an IO like sys.stdout so this can all be combined

File: sssom/cli.py [hunk 5]
```diff
@@ -627,11 +619,7 @@ def rewire(
     g.parse(input, format=input_format)
     rewire_graph(g, msdf, precedence=precedence)
     rdfstr = g.serialize(format=output_format).decode()
-    if output:
-        with open(output, "w") as stream:
-            stream.write(rdfstr)
-    else:
-        print(rdfstr)
+    print(rdfstr, file=output)
```
Comment (cthoyt) [inline sssom/cli.py:622]: print lets you say where to print via the `file` argument, use this to deduplicate code like this a ton

File: sssom/writers.py [hunk 6]
```diff
@@ -59,23 +59,13 @@ def write_table(msdf: MappingSetDataFrame, filename: str, serialisation="tsv") -
     lines = [f"# {line}" for line in lines if line != ""]
     s = msdf.df.to_csv(sep=sep, index=False)
     lines = lines + [s]
-
-    if filename and filename != "-":
```
Comment (cthoyt) [inline sssom/writers.py:63]: ewwwwww all of this is no longer necessary because click does it for you!

File: sssom/writers.py [hunk 7]
```diff
@@ -125,7 +115,8 @@ def write_owl(
         serialisation = SSSOM_DEFAULT_RDF_SERIALISATION
 
     graph = to_owl_graph(msdf)
-    graph.serialize(destination=filename, format=serialisation)
+    t = graph.serialize(format=serialisation, encoding="utf-8")
```
Comment (cthoyt) [inline sssom/writers.py:118]: this is annoying - rdflib only writes to bytes (though I heard they got rid of this in rdflib 6 #138!!!)

File: sssom/writers.py [hunk 8]
```diff
@@ -347,11 +339,11 @@ def get_writer_function(*, output_format: Optional[str] = None, output: str):
 
 
 def write_tables(sssom_dict, output_dir) -> None:
-    for split_id in sssom_dict:
-        sssom_file = os.path.join(output_dir, f"{split_id}.sssom.tsv")
-        msdf = sssom_dict[split_id]
```
Comment (cthoyt) [inline sssom/writers.py:352]: any time you do this in a loop, you can use the `.items()` iterator over the dict instead

**Comment by cthoyt** [comment]
@hrshdhgd same drill as last time, please read through this PR and check out the comments I made about the general philosophy of the changes

File: sssom/cli.py [hunk 9]
```diff
@@ -227,16 +220,14 @@ def ptable(input=None, output=None, inverse_factor=None):
     df = collapse(msdf.df)
     # , priors=list(priors)
     rows = dataframe_to_ptable(df)
-
-    with smart_open(output) as fh:
-        for row in rows:
-            print("\t".join(row), file=fh)
+    for row in rows:
+        print(row, sep="\t", file=output)
```
Comment (cthoyt) [inline sssom/cli.py:224]: note that there's a `sep` option which makes this much more simple

File: sssom/cli.py (hunk 2 above)
Comment (hrshdhgd) [reply sssom/cli.py:324]: Excellent! mypy was complaining about this too! I've never initialized variables to `None` like this ever but again, this is code I inherited so no idea where this style came from.

**Review by hrshdhgd — APPROVED** [review]
Looks good and as usual learnt a few things. Thanks again!

**Pull request merged**